server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java

```diff
@@ -0,0 +1,345 @@
+
+        protected final MultiFields.Builder multiFieldsBuilder = new MultiFields.Builder();
+        protected final CopyTo.Builder copyTo = new CopyTo.Builder();
+
+        /**
+         * Creates a new Builder with a field name
+         */
+        protected Builder(String name) {
+            super(name);
+        }
+
+        /**
+         * Initialises all parameters from an existing mapper
+         */
+        public Builder init(FieldMapper initializer) {
+            for (Parameter<?> param : getParameters()) {
+                param.init(initializer);
+            }
+            for (Mapper subField : initializer.multiFields) {
+                multiFieldsBuilder.add(subField);
+            }
+            return this;
+        }
+
+        private void merge(FieldMapper in, Conflicts conflicts) {
+            for (Parameter<?> param : getParameters()) {
+                param.merge(in, conflicts);
+            }
+            for (Mapper newSubField : in.multiFields) {
+                multiFieldsBuilder.update(newSubField, parentPath(newSubField.name()));
+            }
+            this.copyTo.reset(in.copyTo);
+        }
+
+        /**
+         * @return the list of parameters defined for this mapper
+         */
+        protected abstract List<Parameter<?>> getParameters();
+
+        @Override
+        public abstract ParametrizedFieldMapper build(BuilderContext context);
+
+        /**
+         * Builds the full name of the field, taking into account parent objects
+         */
+        protected String buildFullName(BuilderContext context) {
+            return context.path().pathAsText(name);
+        }
+
+        private void toXContent(XContentBuilder builder, boolean includeDefaults) throws IOException {
+            for (Parameter<?> parameter : getParameters()) {
+                parameter.toXContent(builder, includeDefaults);
+            }
+        }
+
+        /**
+         * Parse mapping parameters from a map of mappings
+         * @param name              the field mapper name
+         * @param parserContext     the parser context
+         * @param fieldNode         the root node of the map of mappings for this field
+         */
+        public final void parse(String name, TypeParser.ParserContext parserContext, Map<String, Object> fieldNode) {
+            Map<String, Parameter<?>> paramsMap = new HashMap<>();
+            for (Parameter<?> param : getParameters()) {
+                paramsMap.put(param.name, param);
+            }
+            String type = (String) fieldNode.remove("type");
+            for (Iterator<Map.Entry<String, Object>> iterator = fieldNode.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                final String propName = entry.getKey();
+                final Object propNode = entry.getValue();
+                if (Objects.equals("fields", propName)) {
```

Comment:
this means that every field supports multi fields and copy_to, I think. But I think that runtime fields won't?

Reply:
We can add disabling feature flags later if we need to.

Reply:
Every field does support multi fields and copy_to right now. I'm happy to juggle things for runtime fields in a follow up.